File: frontend/src/services/competition.ts
import apiClient from './api'

// شکل پاسخ رقابت از سمت بک‌اند (snake_case)
export interface Competition {
  id: number
  slot_id: number
  venue_id: number
  offered_price: number
  status: 'active' | 'won' | 'lost' | 'expired'
  expires_at: string
  created_at: string
}

export const competitionService = {
  // شروع رقابت قیمت برای یک سانس آزاد (فقط مدیر سالن مالک)
  start: async (data: { slotId: number; offeredPrice: number }): Promise<Competition> => {
    const response = await apiClient.post('/competitions/start', {
      slot_id: data.slotId,
      offered_price: data.offeredPrice,
    })
    return response.data
  },

  // ثبت پیشنهاد جدید (باید کمتر از بهترین پیشنهاد فعلی باشد)
  placeBid: async (slotId: number, offeredPrice: number): Promise<Competition> => {
    const response = await apiClient.post(`/competitions/${slotId}/bid`, {
      offered_price: offeredPrice,
    })
    return response.data
  },

  // بهترین (کمترین) پیشنهاد فعلی برای یک سانس
  getBestBid: async (slotId: number): Promise<{ best_price: number | null }> => {
    const response = await apiClient.get(`/competitions/slot/${slotId}/best`)
    return response.data
  },
}

File: backend/app/models/competition.py
from sqlmodel import SQLModel, Field, Relationship
from typing import List, Optional
from datetime import datetime, timezone
from enum import Enum

class CompetitionStatus(str, Enum):
    ACTIVE = "active"
    WON = "won"
    LOST = "lost"
    EXPIRED = "expired"

class PriceCompetition(SQLModel, table=True):
    __tablename__ = "price_competitions"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    slot_id: int = Field(foreign_key="slots.id")
    venue_id: int = Field(foreign_key="venues.id")
    venue_manager_id: int = Field(foreign_key="users.id")
    offered_price: int
    status: CompetitionStatus = Field(default=CompetitionStatus.ACTIVE)
    expires_at: datetime
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    
    slot: "Slot" = Relationship(
        back_populates="competitions",
        sa_relationship_kwargs={"foreign_keys": "[PriceCompetition.slot_id]"}
    )
    venue: "Venue" = Relationship()
    venue_manager: "User" = Relationship(back_populates="competitions")
    winning_slots: List["Slot"] = Relationship(
        back_populates="competition_winner",
        sa_relationship_kwargs={"foreign_keys": "[Slot.competition_winner_id]"}
    )
